src/docs/perf/rendering/shader.md

```diff
@@ -0,0 +1,171 @@
+Such tests can also be used to verify the performance change
+before and after the SkSL warm-up. Even better, you can put
+those tests into a CI (continuous integration) system so the
+SkSLs are generated and tested automatically over the lifetime of an app.
+
+Take [Flutter Gallery][] as an example. The CI system is set up to
+generate SkSLs for every Flutter commit, and verifies the performance,
+in the [`transitions_perf_test.dart`][] test.
+For more details, see the [`flutter_gallery_sksl_warmup__transition_perf`][]
+and [`flutter_gallery_sksl_warmup_ios32__transition_perf`][] tasks.
+
+
+[Flutter Gallery]: {{site.github}}/flutter/flutter/tree/master/dev/integration_tests/flutter_gallery
+[`flutter_gallery_sksl_warmup__transition_perf`]: {{site.github}}/flutter/flutter/blob/master/dev/devicelab/bin/tasks/flutter_gallery_sksl_warmup__transition_perf.dart
+[`flutter_gallery_sksl_warmup_ios32__transition_perf`]: {{site.github}}/flutter/blob/master/dev/devicelab/bin/tasks/flutter_gallery_sksl_warmup_ios32__transition_perf.dart
+[integration tests]: /docs/cookbook/testing/integration/introduction
+[`transitions_perf_test.dart`]: {{site.github}}/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/test_driver/transitions_perf_test.dart
+
+## Limitations and considerations
+
+1. **Why not just compile or warm up _all_ possible shaders?**<br>
+   If there are only a limited number of possible shaders,
+   then Flutter could do a warm-up and compile all of them
+   before-hand to avoid such jank.
+   However, for the best overall performance,
+   the Skia GPU backend used by Flutter can dynamically generate
+   shaders based on a lot of parameters at runtime
+   (for example, draws, device models, and driver versions).
+   Due to all possible combinations of those parameters,
+   the number of possible shaders is exponential.
+   In short, Flutter uses programs (app, Flutter, and Skia code)
+   to generate some other programs (shaders). The number of possible
+   shader programs that Flutter can generate is simply too large.
+
+2. **Can SkSLs captured from one device help shader compilation jank
+   on another device?**<br>
+   Theoretically, there's no guarantee that the SkSLs from one device
+   would help on another device (but they also won't cause any troubles
+   if SkSLs aren't compatible across devices).
+   Practically, as shown in the table on this [SkSL-based warmup PR][],
+   SkSLs work surprisingly well
+   even if 1) SkSLs are captured from iOS and then applied to Android devices,
+   or 2) SkSLs are captured from emulators and then applied to real mobile
+   devices. As the Flutter team has only a limited number of devices in the lab,
+   we currently don't have enough data to provide a big picture of cross-device
+   helpfulness. We'd love you to provide us more data points to see how it
+   works in the wild. ([`FrameTiming`][] can be used to compute the worst frame
+   rasterization time in release mode.)
```

Comment:
Maybe add `And the worst frame rasterization time is a good indicator on how severe the shader compilation jank is.`

Reply:
Done